src/main/java/org/opensearch/neuralsearch/highlight/NeuralHighlighterManager.java

```diff
@@ -0,0 +1,323 @@
+            throw new IllegalArgumentException(String.format(Locale.ROOT, "Field %s is empty", fieldContext.fieldName));
+        }
+        return hitValue;
+    }
+
+    /**
+     * Extracts the original query text from the search query object.
+     *
+     * @param query The query object from which to extract the original text
+     * @param fieldName The name of the field being highlighted
+     * @return The extracted original query text for highlighting
+     * @throws IllegalArgumentException if the extracted query text is empty
+     */
+    public String extractOriginalQuery(Query query, String fieldName) {
+        if (fieldName == null) {
+            log.warn("Field name is null, extraction may be less accurate");
+        }
+        return queryTextExtractorRegistry.extractQueryText(query, fieldName);
+    }
+
+    /**
+     * Gets the model ID from the options
+     *
+     * @param options The options map
+     * @return The model ID
+     */
+    public String getModelId(Map<String, Object> options) {
+        Object modelId = options.get(MODEL_ID_FIELD);
+        if (Objects.isNull(modelId)) {
+            throw new IllegalArgumentException(String.format(Locale.ROOT, "Missing required option: %s", MODEL_ID_FIELD));
+        }
+        return modelId.toString();
+    }
+
+    /**
+     * Gets highlighted text from the ML model
+     *
+     * @param modelId The ID of the model to use
+     * @param question The search query
+     * @param context The document text
+     * @return Formatted text with highlighting
+     */
+    public String getHighlightedSentences(String modelId, String question, String context) {
+        List<Map<String, Object>> result = fetchModelResults(modelId, question, context);
+        if (result == null) {
+            return StringUtils.EMPTY;
+        }
+
+        return applyHighlighting(context, result);
+    }
+
+    /**
+     * Fetches highlighting results from the ML model
+     *
+     * @param modelId The ID of the model to use
+     * @param question The search query
+     * @param context The document text
+     * @return The highlighting results
+     */
+    public List<Map<String, Object>> fetchModelResults(String modelId, String question, String context) {
+
+        CountDownLatch latch = new CountDownLatch(1);
+        AtomicReference<List<Map<String, Object>>> resultRef = new AtomicReference<>();
+        AtomicReference<Exception> exceptionRef = new AtomicReference<>();
+
+        SentenceHighlightingRequest request = SentenceHighlightingRequest.builder()
+            .modelId(modelId)
+            .question(question)
+            .context(context)
+            .build();
+
+        mlCommonsClient.inferenceSentenceHighlighting(request, ActionListener.wrap(result -> {
+            resultRef.set(result);
+            latch.countDown();
+        }, exception -> {
+            exceptionRef.set(exception);
+            latch.countDown();
+        }));
```

Comment:
I don't see this latch pattern anywhere else in OpenSearch. Could you use [PlainActionFuture](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/support/PlainActionFuture.java) to wrap the action listener with a future to make it synchronous instead of managing atomic references with callbacks?

Reply:
Thanks @q-andy for suggestion, updated to use [PlainActionFuture](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/support/PlainActionFuture.java) here.